src/NuGet.Clients/NuGet.CommandLine/Common/Solution.cs

```diff
@@ -113,6 +113,7 @@ private void LoadSolutionWithMsbuild14(Assembly msbuildAssembly, string solution
         {
             var solutionFileType = msbuildAssembly.GetType("Microsoft.Build.Construction.SolutionFile");
             var parseMethod = solutionFileType.GetMethod("Parse", BindingFlags.Static | BindingFlags.Public);
+            var projectShouldBuildMethod = solutionFileType.GetMethod("ProjectShouldBuild", BindingFlags.NonPublic | BindingFlags.Instance);
```

Comment:
Our ParseSolutionFilter method is private, so that wouldn't be better. It looks like you're guaranteeing Parse was called before ProjectShouldBuild. (If it isn't, ProjectShouldBuild may give wrong answers.)

Other than that implicit dependency—which could probably be erased by checking a "parse invoked" bool and calling parse if not—I don't know of any reason it has to be internal, so I think this is fine.

Reply:
> I don't know of any reason it has to be internal, so I think this is fine.

Am I to interpret this as you'll make this public on the MSBuild side?

Reply:
I think that's reasonable, so I made a PR:
https://github.com/dotnet/msbuild/pull/7992

I left it as a draft to gather consensus before committing to it.

Reply:
Thanks for starting that @Forgind!

Reply:
@rainersigwald was not a fan of making ProjectShouldBuild public and further suggested NuGet should avoid relying on SolutionFile.Parse.

I'm going to start an internal thread about this. Stay tuned.

Reply:
After some discussions, we have decided to use the internal method for now.